src/main/scala/org/broadinstitute/hail/driver/ExportVCF.scala

```diff
@@ -205,7 +205,58 @@ object ExportVCF extends Command {
           sb.append("GT:AD:DP:GQ:PL")
         gs.foreach { g =>
           sb += '\t'
-          sb.append(g)
+
+          sb.append(g.gt.map { gt =>
+            val p = Genotype.gtPair(gt)
+            s"${ p.j }/${ p.k }"
+          }.getOrElse("./."))
+
+          val missingField: Array[String] = Array(".")
+
+          def concatAppend(toAppend: Array[_]*): Unit = {
+            sb += ':'
+            toAppend.foreachBetween(_.foreachBetween(sb.append(_))(sb += ','))(sb += ':')
+          }
+
+          (g.ad,
+            g.dp.map(Array(_)),
+            g.gq.map(Array(_)),
+            if (g.isDosage)
+              g.dosage
+            else
+              g.pl) match {
+            case (None, None, None, None) =>
+            case (Some(ad), None, None, None) =>
+              concatAppend(ad)
+            case (None, Some(dp), None, None) =>
+              concatAppend(missingField, dp)
+            case (Some(ad), Some(dp), None, None) =>
+              concatAppend(ad, dp)
+            case (None, None, Some(gq), None) =>
+              concatAppend(missingField, missingField, gq)
+            case (Some(ad), None, Some(gq), None) =>
+              concatAppend(ad, missingField, gq)
+            case (None, Some(dp), Some(gq), None) =>
+              concatAppend(missingField, dp, gq)
+            case (Some(ad), Some(dp), Some(gq), None) =>
+              concatAppend(ad, dp, gq)
+            case (None, None, None, Some(plOrD)) =>
+              concatAppend(missingField, missingField, missingField, plOrD)
+            case (Some(ad), None, None, Some(plOrD)) =>
+              concatAppend(ad, missingField, missingField, plOrD)
+            case (None, Some(dp), None, Some(plOrD)) =>
+              concatAppend(missingField, dp, missingField, plOrD)
+            case (Some(ad), Some(dp), None, Some(plOrD)) =>
+              concatAppend(ad, dp, missingField, plOrD)
+            case (None, None, Some(gq), Some(plOrD)) =>
+              concatAppend(missingField, missingField, gq, plOrD)
+            case (Some(ad), None, Some(gq), Some(plOrD)) =>
+              concatAppend(ad, missingField, gq, plOrD)
+            case (None, Some(dp), Some(gq), Some(plOrD)) =>
+              concatAppend(missingField, dp, gq, plOrD)
+            case (Some(ad), Some(dp), Some(gq), Some(plOrD)) =>
+              concatAppend(ad, dp, gq, plOrD)
+          }
```

Comment:
Here's roughly what I was imagining:

```
(g.ad, g.dp, g.gq,
  if (g.isDosage)
    g.dosage.map(Left)
  else
    g.pl.map(Right)) match {
  case (None, None, None, None) =>
  case (Some(ad), None, None, None) =>
    append(sb, ad) // Array[Int]
  case (ad, Some(dp), None, None) =>
    append(sb, ad) // Option[Array[Int]]
   sb += ':'
    append(sb, dp) // Array[Int]
  case (ad, dp, Some(gq), None) =>
    append(sb, ad)
    sb += ':'
    append(sb, dp) // Option[Array[Int]]
    sb.append(sb, gq) // Int
  case (ad, dp, gq, Some(dosageOrPL) =>
     append ad, dp, gq...
     dosageOrPL match {
       case Left(dosage) => append(sb, dosage)
       case Right(pl) => append(sb, pl)
    }
}
```

I'd put the variants of `append` at the top level and pass in `sb`.  Otherwise, I think Scala implements local definitions with an inner class that allocates.  This is pretty clean and only allocates one Left/Right object per genotype.

What do you think?